### PR TITLE
Deploy to to Kubernetes by using the Aspirate tool

### PR DESCRIPTION
**Pull request by alistairmatthews**
## Summary

A new document that covers deployment to a Kubernetes cluster by using the [Aspirate tool](https://www.nuget.org/packages/Aspirate)

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/deployment/k8s-deployment.md](https://github.com/dotnet/docs-aspire/blob/8c17346bdb05a5ee71e3c3f9444e1aab0ae00b96/docs/deployment/k8s-deployment.md) | [Deploy .NET Aspire apps to a Kubernetes cluster](https://review.learn.microsoft.com/en-us/dotnet/aspire/deployment/k8s-deployment?branch=pr-en-us-134) |

### CODE REVIEW

File: docs/deployment/k8s-deployment.md [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+
+[!INCLUDE [aspire-prereqs](../includes/aspire-prereqs.md)]
+
+## Aspirate prerequisites
+
+The Aspirate tool needs the following prerequisites to be in place before you can use it to deploy a .NET Aspire app:
+
+- .NET 8 installed on the computer where you will run Aspirate. You use the `dotnet` CLI to install Aspirate and the .NET framework is required for the app to be built.
+- A container registry. You can either specify the location of the registry in the Aspire manifest file or configure it when you run Aspirate for the first time. After Aspirate builds the app, it uploads the resulting container images to this registry.
+- The Kuberbetes CLI tool, `kubectl`, installed on the complete where you will run Aspirate.
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed. I linked to the [Install Tools](https://kubernetes.io/docs/tasks/tools/) document.

File: docs/deployment/k8s-deployment.md [hunk 2]
```diff
@@ -0,0 +1,182 @@
...
+- A Kubernetes cluster. The Kubernetes CLI tool must be configured to connect to this cluster in the **kubeconfig** file.
+
+## Comparing manifest formats
+
+Manifest files are used in many application development and hosting platforms to describe the components of an app. Aspire can create manifest files that help tool developers to create  deployment code. Kubernetes also uses manifest files to describe how the system should create and manage resources in the cluster. However, these two types of manifest file have completely different formats and are not interchangeable.
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed throughout document.

File: docs/deployment/k8s-deployment.md [hunk 3]
```diff
@@ -0,0 +1,182 @@
...
+## Aspirate prerequisites
+
+The Aspirate tool needs the following prerequisites to be in place before you can use it to deploy a .NET Aspire app:
+
+- .NET 8 installed on the computer where you will run Aspirate. You use the `dotnet` CLI to install Aspirate and the .NET framework is required for the app to be built.
+- A container registry. You can either specify the location of the registry in the Aspire manifest file or configure it when you run Aspirate for the first time. After Aspirate builds the app, it uploads the resulting container images to this registry.
+- The Kuberbetes CLI tool, `kubectl`, installed on the complete where you will run Aspirate.
+- A Kubernetes cluster. The Kubernetes CLI tool must be configured to connect to this cluster in the **kubeconfig** file.
+
+## Comparing manifest formats
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: FIxed throughout document.

File: docs/deployment/k8s-deployment.md [hunk 4]
```diff
@@ -0,0 +1,182 @@
...
+You can use the `dotnet` command to install Aspirate from NuGet:
+
+```bash
+dotnet tool install -g aspirate --prerelease
+```
+
+> [!NOTE]
+> At the time of writing, the Aspirate tool is in preview and the `--prerelease` option is required.
+
+## Initializing Aspirate
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed.

File: docs/deployment/k8s-deployment.md [hunk 5]
```diff
@@ -0,0 +1,182 @@
...
+- **ContainerTag.** This tag will be added to the Docker images Aspirate creates.
+- **TemplatePath.** This value modifies the default path to Visual Studio template that wil be transformed into manifests.
+
+You can manually set these values in the project file, or issue this command:
+
+```bash
+aspirate init
+```
+
+Aspirate leads you through the process of setting these values. They are persisted in an **aspirate.json** file.
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed.

File: docs/deployment/k8s-deployment.md [hunk 6]
```diff
@@ -0,0 +1,182 @@
...
+
+## Creating manifests
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed.

File: docs/deployment/k8s-deployment.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+You can use the `generate` command to create Kubernetes manifests for the projects in your solution. Before you issue the command, change to the top level folder for your **AppHost** project:
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed.

File: docs/deployment/k8s-deployment.md [hunk 8]
```diff
@@ -0,0 +1,182 @@
...
+
+```bash
+cd AspireApp.AppHost
+aspirate generate
+```
+
+The `generate` command:
+
+- Creates Kubernetes manifest file for each component in the solution and stores them in the **aspirate-output** subfolder.
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed.

File: docs/deployment/k8s-deployment.md [hunk 9]
```diff
@@ -0,0 +1,182 @@
...
+
+## Install containers in a Kubernetes cluster
+
+Once the containers are built and stored in your registry, you can use Aspirate to run them on your Kubernetes cluster:
+
+```bash
+aspirate apply
+```
+
+The command runs the containers on the cluster specified in your **kubeconfig** file.
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed.

**Comment by IEvangelist** [comment]
This is great @alistairmatthews - One more thing, we need a TOC entry for this.

File: docs/deployment/k8s-deployment.md [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+
+Aspirate takes three settings from the project file for the .NET Aspire AppHost project:
+
+- **ContainerRegistry**: The location of a Docker image registry. Aspirate stores the images it creates in this registry and deploys to the Kubernetes cluster from there.
+- **ContainerTag**: Aspirate adds this tag to the Docker images it creates.
+- **TemplatePath**: Modifies the default path to Visual Studio template that Aspirate transforms into manifests.
+
+You can manually set these values in the project file, or issue this command:
+
+```bash
```
Comment (IEvangelist) [inline docs/deployment/k8s-deployment.md:148]:
```suggestion
```dotnetcli
```

File: docs/deployment/k8s-deployment.md [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+aspirate init
+```
+
+Aspirate leads you through the process of setting these values and persists them in an _aspirate.json_ file.
+
+## Create manifests
+
+To create Kubernetes manifests for the projects in your solution, use the `generate` command. Before you issue the command, change to the top-level folder for your _*.AppHost_ project:
+
+```bash
```
Comment (IEvangelist) [inline docs/deployment/k8s-deployment.md:158]:
```suggestion
```dotnetcli
```

File: docs/deployment/k8s-deployment.md [hunk 12]
```diff
@@ -0,0 +1,181 @@
...
+- Pushes the built containers to the **ContainerRegistry** you specified in the `init` command.
+
+> [!NOTE]
+> The `aspirate build` command is similar to the `generate` command. It creates manifest files, builds projects, and pushes the containers to the registry by default. However, if you use the `--aspire-manifest` option, you can use an existing .NET Aspire manifest file.
+
+## Install containers in a Kubernetes cluster
+
+Once the containers are built and stored in your registry, you can use Aspirate to run them on your Kubernetes cluster:
+
+```bash
```
Comment (IEvangelist) [inline docs/deployment/k8s-deployment.md:177]:
```suggestion
```dotnetcli
```

**Review by IEvangelist — COMMENTED** [review]
I think we should use .NET CLI instead of Bash here. This is a global tool, but it's powered by the .NET CLI.

File: docs/deployment/k8s-deployment.md [hunk 13]
```diff
@@ -0,0 +1,181 @@
...
+      terminationGracePeriodSeconds: 180
+```
+
+You can use the Aspirate tool to create the Kubernetes manifest files and containers and deploy them to a cluster.
+
+## Install Aspirate
+
+To install Aspirate, use the `dotnet tool` command:
+
+```bash
```
[the remark this answers is not in the document]
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed

File: docs/deployment/k8s-deployment.md (hunk 11 above)
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed.

File: docs/deployment/k8s-deployment.md (hunk 12 above)
Comment (alistairmatthews) [reply docs/deployment/k8s-deployment.md]: Fixed.

**Comment by alistairmatthews** [comment]
> This is great @alistairmatthews - One more thing, we need a TOC entry for this.

@IEvangelist - I think I've done that. You might like to check my TOC entry to make sure I've got it right. Thanks!

**Comment by davidfowl** [comment]
I think @prom3theu5 should review this. Also there are still changes being made so maybe we need to keep iterating on this??

**Comment by mitchdenny** [comment]
Great to see docs for this!

**Comment by prom3theu5** [comment]
Awesome - thanks 😄 

Will track the link and as and when changes are implemented update accordingly

**Pull request merged**